samza-core/src/main/java/org/apache/samza/classloader/IsolatingClassLoaderFactory.java

```diff
@@ -0,0 +1,195 @@
+      ClassLoader apiClassLoader) {
+    return LoaderBuilder.anIsolatingLoader()
+        // look in application directory for JARs
+        .withClasspath(getClasspathAsURIs(applicationLibDirectory))
+        // getClasspathAsURIs should already satisfy this, but doing it to be safe
+        .withOriginRestriction(OriginRestriction.denyByDefault().allowingDirectory(applicationLibDirectory, false))
+        .withParentRelationship(buildApiParentRelationship(apiLibDirectory, apiClassLoader))
+        .build();
+  }
+
+  private static ClassLoader buildInfrastructureClassLoader(File infrastructureLibDirectory, File apiLibDirectory,
+      ClassLoader apiClassLoader, ClassLoader applicationClassLoader) {
+    return LoaderBuilder.anIsolatingLoader()
+        // look in application directory for JARs
+        .withClasspath(getClasspathAsURIs(infrastructureLibDirectory))
+        // getClasspathAsURIs should already satisfy this, but doing it to be safe
+        .withOriginRestriction(OriginRestriction.denyByDefault().allowingDirectory(infrastructureLibDirectory, false))
+        .withParentRelationship(buildApiParentRelationship(apiLibDirectory, apiClassLoader))
+        // need to be able to fall back to application for loading certain classes
+        .addFallbackDelegate(buildFallbackDelegateRelationship(applicationClassLoader))
+        .build();
+  }
+
+  private static DelegateRelationship buildApiParentRelationship(File apiLibDirectory, ClassLoader apiClassLoader) {
+    DelegateRelationshipBuilder apiParentRelationshipBuilder = DelegateRelationshipBuilder.builder()
+        // needs to load API classes from the API classloader
+        .withDelegateClassLoader(apiClassLoader)
+        // use FULL to only load API classes from API classloader
+        .withIsolationLevel(IsolationLevel.FULL);
+    /*
+     * All bootstrap classes (e.g. java.lang classes) should be accessible and loaded from a single classloader, so that
+     * they are the same across the whole JVM. The API classloader has the bootstrap classes.
+     */
+    apiParentRelationshipBuilder.addDelegatePreferredClassPredicate(new BootstrapClassPredicate());
+    // add the classes which are API classes
+    getApiClasses(apiLibDirectory).forEach(
+        apiClassName -> apiParentRelationshipBuilder.addDelegatePreferredClassPredicate(new GlobMatcher(apiClassName)));
+    return apiParentRelationshipBuilder.build();
+  }
+
+  private static DelegateRelationship buildFallbackDelegateRelationship(ClassLoader fallbackClassLoader) {
+    return DelegateRelationshipBuilder.builder()
+        .withDelegateClassLoader(fallbackClassLoader)
+        // want to load any class from the fallback
+        .withIsolationLevel(IsolationLevel.NONE)
+        .build();
+  }
+
+  @VisibleForTesting
+  static List<String> getApiClasses(File directoryWithClassList) {
+    File parentPreferredFile = new File(directoryWithClassList, "application-master-samza-api-class-list.txt");
```

Comment:
What is this file name? Maybe extract constant (to IsolationUtil) and document how/when this is generated.